### PR TITLE
Add compile_error!

### LINKED ISSUE #40872: Tracking issue for the addition of the compile_error! macro
https://github.com/rust-lang/rfcs/pull/1695

### PR DESCRIPTION
**Pull request by wesleywiser**
Related to #40872

### CODE REVIEW

File: src/test/compile-fail/compile_fail_macro.rs [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//#![feature(compile_error_macro)]
```
Comment (wesleywiser) [inline src/test/compile-fail/compile_fail_macro.rs]: This test passes even though the feature flag is commented out. Are macros insta-stable?

Comment (est31) [inline src/test/compile-fail/compile_fail_macro.rs]: > Are macros insta-stable?

Yes, right now they are.

**Comment by arielb1** [comment]
Thanks for the PR @wesleywiser! @brson (or someone else) will be reviewing your PR soon.

File: src/libcore/macros.rs [hunk 2]
```diff
@@ -562,6 +562,17 @@ macro_rules! unimplemented {
 ///
 /// For more information, see documentation for `std`'s macros.
 mod builtin {
+
+    /// Unconditionally causes compilation to fail with the given error message when encountered.
+    ///
+    /// For more information, see the [RFC].
+    ///
+    /// [RFC]: https://github.com/rust-lang/rfcs/blob/master/text/1695-add-error-macro.md
+    #[stable(feature = "compile_error_macro", since="1.19.0")]
```
Comment (est31) [inline src/libcore/macros.rs]: You should have `1.20.0` here.

Comment (wesleywiser) [inline src/libcore/macros.rs]: Thanks

**Comment by ollie27** [comment]
This doesn't have to be insta-stable, see https://github.com/rust-lang/rust/blob/master/src/libsyntax_ext/concat_idents.rs for an example of an unstable macro.

**Comment by est31** [comment]
Oh right, I remembered wrongly, what I thought was deprecation warnings of macros...

**Comment by wesleywiser** [comment]
Quick status update: I'm still working on this. I think the code changes are finished and tomorrow I'll work on updating the unstable book.

**Comment by wesleywiser** [comment]
Pushed unstable docs

**Comment by wesleywiser** [comment]
This should be ready for review.

**Comment by wesleywiser** [comment]
Rebased

**Pull request merged**